makepdf.sh

```diff
@@ -110,30 +146,143 @@ is_pandoc_version_new()
+                         --metadata-file $OLDPWD/pandoc/config-web.yaml \
+                         -o $out_rel_path_file $input_basename".md")
+        
+        else
+                echo "ERROR: can only process one of these styles: opensolaris or web"
+                exit 1
+        fi
+}
+
+
 do_conversion()
 {
-        input=$1
-        output=$2
-        format=$3
-        style=$4
+        input_basename="$1"
+        format="$2"
+        style="$3"
 
+        
         if [ "$format" == "epub" ]; then
-                $(pandoc -t epub --toc -f markdown+grid_tables+table_captions -o $output $input  --pdf-engine=xelatex)
-        elif [ "$format" == "pdf" ]; then
+                do_conversion_epub "$input_basename" "$format"
+
+                
+        elif [ "$format" == "pdf" ]; then               
                 if [ "$(is_pandoc_version_new)" == "false" ]; then
-                        $(pandoc --toc -f markdown+grid_tables+table_captions -o $output $input  --pdf-engine=xelatex)
+                        # This call for backward compatibility 
+                        do_conversion_with_older_pandoc "$input_basename" "$format"
+                        
                 else
-                        if [ "$style" == "opensolaris" ]; then
-                            $(pandoc --pdf-engine=xelatex --lua-filter $OLDPWD/pandoc/filter-sb.lua --metadata-file $OLDPWD/pandoc/config-sb.yaml -o $output $input)
-                        else
-                            $(pandoc --pdf-engine=xelatex --lua-filter $OLDPWD/pandoc/filter-web.lua --metadata-file $OLDPWD/pandoc/config-web.yaml -o $output $input)
-                        fi
+                        do_conversion_pdf "$input_basename" "$format" "$style"
                 fi
         fi
+}
+
 
+# This function sets is_format_valid_result = True || False
+# True: users select a valid format
+# False: users select an invalid format
+is_format_valid()
+{
+        input_format=$1
+        is_format_valid_result="False"
+        
+        for  i in $SUPPORTED_FORMATS  ; do
+                if [ "${i}" == "$input_format" ]; then
+                        is_format_valid_result="True"
+                fi
+        done
 }
 
 
+make_out_dir()
+{
+        the_output_format=$1
+        path_to_input=$2
+        
+        is_format_valid $the_output_format
+        if [ "$is_format_valid_result" == "False" ]; then
+                echo >&2 "ERROR: -t only supports the following formats: "
+                for  i in $SUPPORTED_FORMATS  ; do
+                        echo "    "$i
+                done
+                exit 1
+        fi
+        
+        out_path=$(pwd)"/"$path_to_input"/${the_output_format}/"
+        mkdir -p $out_path
+        if [ ! -d "$out_path" ]; then
+                echo >&2 "ERROR: cannot create director at: "$out_path
+                exit 1
+        fi
+
+        echo "  Writing output to this directory: " $out_path
```

Comment:
The jenkins build server probably does not require any output except possibly to stderr. So possibly silent all output by default?   I think this is the possibly the more appropriate remedy. '/pdf/contrib', better is [whatever the option to -t, default is pdf]/contrib. So who uses this information? Probably the developer, i.e., me. Maybe a -v option for verbosity, and silence all output except stderr, unless -v is oused is probably the bettgter way to go.

Reply:
Could do that, I don't have any preference. In the case of the CI, we'll want to know what file was being processed if there is an error from Pandoc. I'm not sure if Pandoc consistently gives the filename in it's error output.
One thing the script doesn't do which the CI would benefit from is aborting when there is a Pandoc error. Now I think about it, it may be better to have an entirely separate script just for the CI to use - although that would increase complexity from a maintenance perspective.

Reply:
As stated above, I think the script does catch pandoc errors; assuming pandoc reports the error; although my impression is that it has a lot of general exceptions, leaving me often to guess what actually happened.

It is Bash, which means it is horrible to program in. Better would be Python, but Python is not always available and there was that awful change from 2.7 to 3.x, which OI was caught up with when I originally wrote the thing

Reply:
I responded in the other comment, but exit on error is now working.